x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/CcrRetentionLeaseIT.java

```diff
@@ -0,0 +1,414 @@
+
+        final Map<String, String> additionalSettings = new HashMap<>();
+        additionalSettings.put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), "true");
+        additionalSettings.put(IndexService.RETENTION_LEASE_SYNC_INTERVAL_SETTING.getKey(), TimeValue.timeValueMillis(200).getStringRep());
+        final String leaderIndexSettings = getIndexSettings(numberOfShards, numberOfReplicas, additionalSettings);
+        assertAcked(leaderClient().admin().indices().prepareCreate(leaderIndex).setSource(leaderIndexSettings, XContentType.JSON));
+        ensureLeaderGreen(leaderIndex);
+
+        logger.info("indexing [{}] docs", numberOfDocuments);
+        for (int i = 0; i < numberOfDocuments; i++) {
+            final String source = String.format(Locale.ROOT, "{\"f\":%d}", i);
+            leaderClient().prepareIndex(leaderIndex, "doc", Integer.toString(i)).setSource(source, XContentType.JSON).get();
+            if (rarely()) {
+                leaderClient().admin().indices().prepareFlush(leaderIndex).setForce(true).setWaitIfOngoing(true).get();
+            }
+        }
+
+        leaderClient().admin().indices().prepareFlush(leaderIndex).setForce(true).setWaitIfOngoing(true).get();
+
+        final Settings.Builder settingsBuilder = Settings.builder()
+                .put(IndexMetaData.SETTING_INDEX_PROVIDED_NAME, followerIndex)
+                .put(CcrRepository.RETENTION_LEASE_RENEW_INTERVAL_SETTING.getKey(), TimeValue.timeValueMillis(200))
+                .put(CcrSettings.CCR_FOLLOWING_INDEX_SETTING.getKey(), true);
+        return new RestoreSnapshotRequest(leaderClusterRepoName, CcrRepository.LATEST)
+                .indexSettings(settingsBuilder)
+                .indices(leaderIndex)
+                .indicesOptions(indicesOptions)
+                .renamePattern("^(.*)$")
+                .renameReplacement(followerIndex)
+                .masterNodeTimeout(new TimeValue(1L, TimeUnit.HOURS));
+    }
+
+    public void testRetentionLeaseIsTakenAtTheStartOfRecovery() throws Exception {
+        final String leaderIndex = "leader";
+        final int numberOfShards = randomIntBetween(1, 3);
+        final int numberOfReplicas = between(0, 1);
+        final String followerIndex = "follower";
+        final int numberOfDocuments = scaledRandomIntBetween(1, 8192);
+        final RestoreSnapshotRequest restoreRequest =
+                setUpRestoreSnapshotRequest(leaderIndex, numberOfShards, numberOfReplicas, followerIndex, numberOfDocuments);
+        final RestoreService restoreService = getFollowerCluster().getCurrentMasterNodeInstance(RestoreService.class);
+        final ClusterService clusterService = getFollowerCluster().getCurrentMasterNodeInstance(ClusterService.class);
+
+        final PlainActionFuture<RestoreInfo> future = PlainActionFuture.newFuture();
+        restoreService.restoreSnapshot(restoreRequest, waitForRestore(clusterService, future));
```

Comment:
do we need to access the low-level RestoreService in this test? Can we use a regular putFollow and then check the same stuff as is done further down below in this test?

(same for the other tests in this class)

Reply:
Yes, as this test is about the restore process. We don't want following to start in these tests because following will also be manipulating retention leases, so the test could pass spuriously if the restore process does not manipulate the retention leases in the right way, but later after the restore is complete the follower does (these are assert busies after all).